std/experimental/typecons.d

```diff
@@ -0,0 +1,731 @@
+{
+    @trusted
+    T dynamicCast(S)(inout S source)
+    if (is(S == class) || is(S == interface))
+    {
+        static if (is(Unqual!S : Unqual!T))
+        {
+            import std.traits : QualifierOf;
+            alias Qual = QualifierOf!S; // SharedOf or MutableOf
+            alias TmpT = Qual!(Unqual!T);
+            inout(TmpT) tmp = source;   // bypass opCast by implicit conversion
+            return *cast(T*)(&tmp);     // + variable pointer cast + dereference
+        }
+        else
+        {
+            return cast(T)typecons_d_toObject(*cast(void**)(&source));
+        }
+    }
+}
+
+unittest
+{
+    class C { @disable opCast(T)() {} }
+    auto c = new C;
+    static assert(!__traits(compiles, cast(Object)c));
+    auto o = dynamicCast!Object(c);
+    assert(c is o);
+
+    interface I { @disable opCast(T)() {} Object instance(); }
+    interface J { @disable opCast(T)() {} Object instance(); }
+    class D : I, J { Object instance() { return this; } }
+    I i = new D();
+    static assert(!__traits(compiles, cast(J)i));
+    J j = dynamicCast!J(i);
+    assert(i.instance() is j.instance());
+}
+
+/*
+ * Determines if the `Source` type satisfies all interface requirements of
+ * `Targets`.
+ */
+private template implementsInterface(Source, Targets...)
+if (Targets.length >= 1 && allSatisfy!(isMutable, Targets))
+{
+    import std.meta : staticMap;
+
+    // strict upcast
+    bool implementsInterface()() @safe pure nothrow
+    if (Targets.length == 1 && is(Source : Targets[0]))
+    {
+        return true;
+    }
+    // structural upcast
+    template implementsInterface()
+    if (!allSatisfy!(Bind!(isImplicitlyConvertible, Source), Targets))
+    {
+        auto implementsInterface()
+        {
+            return hasRequiredMethods!();
+        }
+
+        // list of FuncInfo
+        alias TargetMembers = UniqMembers!(ConcatInterfaceMembers!Targets);
+        // list of function symbols
+        alias SourceMembers = GetOverloadedMethods!Source;
+
+        // Check whether all of SourceMembers satisfy covariance target in
+        // TargetMembers
+        template hasRequiredMethods(size_t i = 0)
+        {
+            static if (i >= TargetMembers.length)
+                enum hasRequiredMethods = true;
+            else
+            {
+                enum foundFunc = findCovariantFunction!(TargetMembers[i], Source, SourceMembers);
+                static if (foundFunc == -1)
+                    pragma(msg, "Could not locate matching function for: " ~ TargetMembers[i].stringof);
+                enum hasRequiredMethods =
+                    foundFunc != -1 &&
+                    hasRequiredMethods!(i + 1);
+            }
+        }
+    }
+}
+/// ditto
+private template implementsInterface(Source, Targets...)
+if (Targets.length >= 1 && !allSatisfy!(isMutable, Targets))
+{
+    import std.meta : staticMap;
+
+    alias implementsInterface = .implementsInterface!(Source, staticMap!(Unqual, Targets));
```

Comment:
Why define this instead of letting the user unqualify the interface(s)? Looks like the library is taking too much initiative here. A struct may implement the unqualified but not the qualified version of an interface.